text/0123-improve-span-status-api.md

```diff
@@ -0,0 +1,303 @@
+example a function that calls into other lower level functions) and each of
+those sub-operations might have a notion of status or error. If those
+sub-operations aren't represented with a child span then it is out of scope
+how their status is captured. 
+
+## Explanation
+
+#### Data representation of Span Status
+
+We should aim for representations that require minimal-to-no end-developer
+translation effort from the native representation and capture at least the key
+numeric/string data developers are likely to search for or relate to. I suggest
+status is this set of information:
+
+1. StatusType - The name of the type of error such as "HTTP", "gRPC",
+"LanguageException", "HRESULT", "POSIX", or "ErrorMessage". The list is
+end-user-extensible but common status type names should be standardized.
+(Perhaps there is already some standardization we could borrow?)
+2. StatusData - A discriminated union of:
+   - An integer, a string, or a tuple of integer and string. These options can
+be used for:
+     - Enumerated status codes: For example an http status code could be
+represented as 404, "Not Found", or both. In the case of common status codes
+OpenTelemetry SDK or backend could optionally assist in filling out the
+remainder of a partially specified enumeration value. For enumerations that
+aren't well-known the community of users is responsible for determining any
+conventions.
+     - Free-form error messages
+   - Exception object - whatever the SDK language's default exception datatype
+is, if it has one.
+   - void
+3. SuccessHint - An optional boolean that represents the span author's best
+guess whether this status represents a successful or failed operation, however
+they choose to define those terms. For well-known status types I'd suggest the
+hint be ignored but for user-defined status types this is likely the only clue
+whether the span should be surfaced in a UI as being abnormal or failed.
+
+Although UI creators are free experiment with how the data is presented I
+expect most presentations would either be the StatusData alone, or the
+StatusData qualified with the StatusType and some separator character. For
+example StatusData alone might create names like "FileNotFoundException",
+"503", "E_FAIL (0x80004005)", "SyntaxError on line 405: Did you forget a
+semicolon?", and "BadQuery".
+
+Exceptions could have progressive level of detail drilling into messages,
+stack traces, inner exceptions, links to source, etc if the exporter serialized
+sufficient data but how and whether that occurs is out-of-scope in this design. 
+
+#### API to capture this data
+
+I would suggest an API called Span.SetStatus(...) that takes all the arguments
+above, and optionally overloads or default parameters that make common calls
+easier. For example in C#:
+
+````C#
+void SetStatus(Exception spanException, string statusType = "LanguageException", bool? successHint = null);
```

Comment:
Multiple exceptions can occur during the lifetime of a span and experiencing an exception may or may not result in a non-ok status depending on the circumstances. How can multiple exceptions can handled using this API?

Reply:
Multiple functions returning error codes may be called during the lifetime of a span and an error return may result in a non-ok status or not.

I think it's up to the user/instrumentation to decide which exception was the semantically important one. Most likely the one that went uncaugth.

Reply:
Agreed with @Oberon00. I also attempted to capture this concept in the scoping section above but perhaps the wording wasn't good?

> The operation represented by a span might be composed of sub-operations (for
> example a function that calls into other lower level functions) and each of
> those sub-operations might have a notion of status or error. If those
> sub-operations aren't represented with a child span then it is out of scope
> how their status is captured.